rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+            {}
+          ],
+          "additionalProperties": false
+        }
+      }
+    }
+  },
+  "additionalProperties": false,
+  "required": ["kind", "$id", "name", "properties"],
+  "$defs": {
+    "propertyTypeObject": {
+      "type": "object",
+      "properties": {
+        "type": {
+          "const": "object"
+        },
+        "properties": {
+          "type": "object",
+          "propertyNames": {
+            "$comment": "Property names must be a valid URI to a Property Type",
+            "type": "string",
+            "format": "uri"
+          },
+          "patternProperties": {
+            ".*": {
+              "oneOf": [
+                {
+                  "$ref": "#/$defs/propertyTypeReference"
+                },
+                {
+                  "type": "object",
+                  "properties": {
+                    "type": {
+                      "const": "array"
+                    },
+                    "items": {
+                      "$ref": "#/$defs/propertyTypeReference"
+                    },
+                    "minItems": {
+                      "type": "integer",
+                      "minimum": 0
+                    },
+                    "maxItems": {
+                      "type": "integer",
+                      "minimum": 0
+                    }
+                  },
+                  "required": ["type", "items"],
+                  "additionalProperties": false
+                }
+              ]
+            }
+          },
+          "minimumProperties": 1
+        },
+        "required": {
+          "type": "array",
+          "items": { 
+            "type": "string",
+            "format": "uri"
+          }
+        }
+      },
+      "required": ["type", "properties"],
+      "additionalProperties": false
+    },
+    "propertyTypeReference": {
+      "type": "object",
+      "properties": {
+        "$ref": {
+          "$comment": "Property Object values must be defined through references to the same valid URI to a Property Type",
+          "type": "string",
+          "format": "uri"
+        }
+      },
+      "required": ["$ref"],
+      "additionalProperties": false
+    }
+  }
+}
+```
+
+### Defining an Entity Type's Properties
+
+#### Specifying an Entity Type has Property Types
+
+The data _within_ an Entity is described simply through a set of Property Types
+
+**Example 1**
+
+The `Book` Entity Type could contain the Property Types `Name`, a `Published On` and a `Blurb`, where `Name` is **required**.
+
+Assuming that:
+
+- the `Name`, `Published On`, and `Blurb` Property Types all have values that are instances of the `Text` Data Type
+
+```json
+{
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/types/@alice/entity-type/book",
+  "type": "object",
+  "name": "Book",
+  "properties": {
+    "https://blockprotocol.org/types/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/types/@alice/property-type/name"
```

Comment:
We've been trying to figure that out as well, I'm not sure if there's a way we can constrain that in JSON schema itself, it is unresolved question 3.

If we can't do that then perhaps we can establish our own vocabulary for `$ref` which tends to be how JSON schema adds semantic meaning to keywords, the last fallback is we should just add a paragraph somewhere expressing its necessity and saying that validator implementations should check it.

Reply:
I think we need it to be in any case, don't we? Otherwise you cannot determine from data instances alone which property they refer to (since the keys can differ).

i.e. we should have a paragraph expressing its necessity now, which notes that we'd ideally have this validated automatically. 

It's a pretty fundamental requirement of realising the benefits of the type system, regardless of how we end up enforcing it.

Reply:
Sorry I didn't mean to imply that we don't need to specify this, we do! I was just saying it'd be good if we could encapsulate it more rigorously than a paragraph. However in lieu of a better solution we shall do so now :)

Reply:
Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/9108187fc5d6c497330a677cf22c37d4f381cb74, I've added it to the JSON Schema additions sections rather than adding even more information around the meta schemas.